File: pset2/readability/readability.c
/**
 * CS50x 2021
 * Problem Set 2: readability.c
 */

#include <stdio.h>
#include <cs50.h>
#include <stdbool.h>
#include <math.h>


int main(void)
{
    int coleman_Liau_index_formula(string text);
    void print_grade(int grade);

    int grade;
    string text;

    text = get_string("Text: ");
    grade = coleman_Liau_index_formula(text);
    print_grade(grade);
}


int coleman_Liau_index_formula(string text)
{
    int formula(int letters, int words, int sentences);
    bool is_alpha_num(char c);
    bool is_space(char c);
    bool is_end(char c);

    int i = 0, letters = 0, words = 0, sentences = 0;
    bool is_word;

    while (text[i]) {
        char character = text[i];

        if (is_alpha_num(character)) {
            letters++;
            is_word = true;
        } else if (character == '-' && is_word) {
            is_word = true;
        } else if ((is_space(character) || (character == ',')) && is_word) {
            words++;
            is_word = false;
        } else if (is_end(character) && is_word) {
            words++;
            is_word = false;
            sentences++;
        }

        i++;
    }

    return formula(letters, words, sentences);
}


void print_grade(int grade)
{
    if (grade < 1)
        printf("Before Grade 1\n");
    else if (grade >= 16)
        printf("Grade 16+\n");
    else
        printf("Grade %i\n", grade);
}

bool is_alpha_num(char c)
{
    return ((c >= 'a') && (c <= 'z')) ? true :
           ((c >= 'A') && (c <= 'Z')) ? true :
           ((c >= '0') && (c <= '9')) ? true : false;
}

bool is_space(char c)
{
    return c == ' ' ? true : false;
}

bool is_end(char c)
{
    return c == '.' ? true :
           c == '!' ? true :
           c == '?' ? true : false;
}


int formula(int letters, int words, int sentences)
{
    float L, S;

    // Average number of letters per 100 words.
    L = (float) letters / words * 100;

    // Average number of sentences per 100 words .
    S = (float) sentences / words * 100;

    return (int) round(0.0588 * L - 0.296 * S - 15.8);
}

File: week6/pset6/readability/readability.py
from cs50 import get_string


def get_text():
    """Prompt user for text. If user enter empty string,
    then prompt again."""
    while True:
        string = get_string("Text: ")
        if string != "":
            return string


def text_parser(text):
    """Helper function for parse input text."""
    letters = count_letters(text)
    words = count_words(text)
    sentences = count_sentences(text)

    return letters, words, sentences


def count_letters(text):
    """Count the number of letters from input text."""
    return len([c for c in text if c.isalnum()])


def count_words(text):
    """Count the number of words from input text."""
    return len([w for w in text.split(' ') if w != ""])


def count_sentences(text):
    """Count the number of sentences from input text."""
    return len([w for w in text.replace('? ', '. ').split('.') if w != ""])


def coleman_liau_index(letters, words, sentences):
    """Coleman-Liau index formula. where L is the average number of letters
    per 100 words in the text, and S is the average number of sentences per
    100 words in the text."""
    L = float(letters) / words * 100.0
    S = float(sentences) / words * 100.0

    return round(0.0588 * L - 0.296 * S - 15.8)


def print_grade(grade):
    if grade >= 16:
        print("Grade 16+")
    elif grade < 1:
        print("Before Grade 1")
    else:
        print(f"Grade {grade}")


def main():
    text = get_text()
    letters, words, sentences = text_parser(text)
    grade = coleman_liau_index(letters, words, sentences)
    print_grade(grade)


if __name__ == "__main__":
    main()